File: count-ways-to-build-good-strings/count-ways-to-build-good-strings.java
class Solution {
    int mod = (int)1e9+7;
    long solve(int low, int high, int zero, int one, int curr, long[] dp) {
        if(curr > high) return 0;
        if(dp[curr] != -1) return dp[curr];
        long ans = 0, t = 0;
        if(curr >= low) t = 1;
        ans = t + solve(low, high, zero, one, curr+zero, dp) + solve(low, high, zero, one, curr+one, dp);
        return dp[curr] = (ans)%mod;
    }
    public int countGoodStrings(int low, int high, int zero, int one) {
        long[] dp = new long[high+1];
        Arrays.fill(dp, -1);
        return (int)(solve(low, high, zero, one, 0, dp) % mod);
    }
}

File: count-ways-to-build-good-strings/count-ways-to-build-good-strings.py
class Solution:
    def countGoodStrings(self, low, high, zero, one):
        mod = int(1e9) + 7
        dp = [-1] * (high + 1)
        return (self.solve(low, high, zero, one, 0, dp) // 2) % mod

    def solve(self, low, high, zero, one, curr, dp):
        if curr > high:
            return 0
        if dp[curr] != -1:
            return dp[curr]
        ans, t = 0, 0
        if curr >= low:
            t = 2
        ans = t + self.solve(low, high, zero, one, curr + zero, dp) + self.solve(low, high, zero, one, curr + one, dp)
        dp[curr] = ans
        return dp[curr]
